### PR TITLE
Simplify wait in slices logic. NFC.

### PR DESCRIPTION
**Pull request by kleisauke**
Split from PR #13007.

### CODE REVIEW

File: system/lib/libc/musl/src/thread/pthread_barrier_wait.c [hunk 1]
```diff
@@ -88,17 +88,16 @@ int pthread_barrier_wait(pthread_barrier_t *b)
 			a_spin();
 		a_inc(&inst->finished);
 #ifdef __EMSCRIPTEN__
-		int is_runtime_thread = emscripten_is_main_runtime_thread();
+		const int is_runtime_thread = emscripten_is_main_runtime_thread();
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_barrier_wait.c:91]: I think think this make any difference for basic types like int.

File: system/lib/pthread/library_pthread.c [hunk 2]
```diff
@@ -85,33 +85,29 @@ void emscripten_thread_sleep(double msecs) {
   double now = emscripten_get_now();
   double target = now + msecs;
 
-  __pthread_testcancel(); // pthreads spec: sleep is a cancellation point, so must test if this
-                          // thread is cancelled during the sleep.
-  emscripten_current_thread_process_queued_calls();
-
   // If we have less than this many msecs left to wait, busy spin that instead.
-  const double minimumTimeSliceToSleep = 0.1;
+  const double minTimeSliceToSleep = 0.1;
```
Comment (sbc100) [inline system/lib/pthread/library_pthread.c:89]: Lets use snake case since we are in C and drop the `const`.   e.g. `double min_time_slice_to_sleep`.    Lets make  that into its own cleanup PR?

File: system/lib/pthread/library_pthread.c [hunk 3]
```diff
@@ -640,7 +634,7 @@ void emscripten_current_thread_process_queued_calls() {
   pthread_mutex_unlock(&call_queue_lock);
 
   // If the queue was full and we had waiters pending to get to put data to queue, wake them up.
-  emscripten_futex_wake((void*)&q->call_queue_head, 0x7FFFFFFF);
+  emscripten_futex_wake((void*)&q->call_queue_head, INT_MAX);
```
Comment (sbc100) [inline system/lib/pthread/library_pthread.c:637]: This can be part of the cleanup PR too.

File: system/lib/pthread/library_pthread.c [hunk 4]
```diff
@@ -386,24 +382,22 @@ static CallQueue* GetOrAllocateQueue(void* target) {
...
+
+  double timeoutUntilTime = emscripten_get_now() + timeoutMSecs;
+  do {
+    emscripten_futex_wait(&call->operationDone, 0, timeoutMSecs);
+    done = atomic_load(&call->operationDone);
+
+    timeoutMSecs = timeoutUntilTime - emscripten_get_now();
+  } while (!done && timeoutMSecs > 0);
+
+  emscripten_set_current_thread_status(EM_THREAD_STATUS_RUNNING);
```
Comment (sbc100) [inline system/lib/pthread/library_pthread.c:398]: Is the something else going on here other than some re-ordering of the code?

File: system/lib/pthread/library_pthread.c [hunk 5]
```diff
@@ -386,24 +382,22 @@ static CallQueue* GetOrAllocateQueue(void* target) {
...
+
+  return done ? EMSCRIPTEN_RESULT_SUCCESS : EMSCRIPTEN_RESULT_TIMED_OUT;
```
Comment (sbc100) [inline system/lib/pthread/library_pthread.c:400]: This line can go in the cleanup PR too.

File: system/lib/libc/musl/src/thread/__timedwait.c [hunk 6]
```diff
@@ -95,11 +94,13 @@ int __timedwait_cp(volatile int *addr, int val,
 	r = -__futex4_cp(addr, FUTEX_WAIT|priv, val, top);
 #endif
 	if (r != EINTR && r != ETIMEDOUT && r != ECANCELED) r = 0;
+#ifndef __EMSCRIPTEN__ // XXX Emscripten revert musl commit a63c0104e496f7ba78b64be3cd299b41e8cd427f
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/__timedwait.c:97]: What is this change doing?   Can/should it split into its own PR?

File: system/lib/libc/musl/src/thread/__timedwait.c [hunk 7]
```diff
@@ -33,18 +32,20 @@ static int __futex4_cp(volatile void *addr, int op, int val, const struct timesp
...
+#endif
 
 int __timedwait_cp(volatile int *addr, int val,
 	clockid_t clk, const struct timespec *at, int priv)
 {
 	int r;
 	struct timespec to, *top=0;
 
+#ifndef __EMSCRIPTEN__
 	if (priv) priv = FUTEX_PRIVATE;
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/__timedwait.c:47]: Why this?

File: system/lib/libc/musl/src/thread/__timedwait.c [hunk 8]
```diff
@@ -57,36 +58,34 @@ int __timedwait_cp(volatile int *addr, int val,
...
-
 #ifdef __EMSCRIPTEN__
-	double msecsToSleep = top ? (top->tv_sec * 1000 + top->tv_nsec / 1000000.0) : INFINITY;
-	int is_runtime_thread = emscripten_is_main_runtime_thread();
+	pthread_t self = __pthread_self();
+	double msecsToSleep = top ? (top->tv_sec * 1000.0 + top->tv_nsec / 1000000.0) : INFINITY;
+	const int is_runtime_thread = emscripten_is_main_runtime_thread();
+
+	// Main runtime thread may need to run proxied calls, so sleep in very small slices to be responsive.
+	const double maxMsecsToSleep = is_runtime_thread ? 1 : 100;
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/__timedwait.c:67]: Split out hoisting of `self` into a separate PR?   (combine with use of `self` in 
`system/lib/libc/musl/src/thread/__wait.c`)

File: system/lib/libc/musl/src/thread/__wait.c [hunk 9]
```diff
@@ -8,29 +8,34 @@
...
 		else return;
 	}
 	if (waiters) a_inc(waiters);
 #ifdef __EMSCRIPTEN__
-	int is_runtime_thread = emscripten_is_main_runtime_thread();
+	pthread_t self = __pthread_self();
+	const int is_runtime_thread = emscripten_is_main_runtime_thread();
+
+	// Main runtime thread may need to run proxied calls, so sleep in very small slices to be responsive.
+	const double maxMsecsToSleep = is_runtime_thread ? 1 : 100;
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/__wait.c:24]: Use snake_case in C code.

File: system/lib/libc/musl/src/thread/pthread_barrier_wait.c (hunk 1 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_barrier_wait.c]: Ah yes, reverted with commit 4de581ac1ba86a6937974736fc1d96cc9aebb447.

File: system/lib/pthread/library_pthread.c (hunk 2 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: I'll split that into its own PR. Although, this file contains many camelCase variables. Maybe we should do that only for the code we touch in musl?

File: system/lib/pthread/library_pthread.c (hunk 3 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: Reverted with commit 165078536c1470685ac7a76db7868d5d601cdbf0, I'll make a PR with these changes only.

File: system/lib/pthread/library_pthread.c (hunk 4 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: Consolidation of the if-statements makes this diff difficult to read. Reverted with commit d78e59fbcf314f573973bf58b52e314aeb27ad4f for now.

File: system/lib/pthread/library_pthread.c (hunk 5 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: Reverted with commit d78e59fbcf314f573973bf58b52e314aeb27ad4f for now, I'll make a PR later with this change.

File: system/lib/libc/musl/src/thread/__timedwait.c (hunk 6 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/__timedwait.c]: It was excluding code that is not related to Emscripten. Reverted with commit 6ea3ad394b1b870f157be7162b7546f64a8896e3 for now.

File: system/lib/libc/musl/src/thread/__timedwait.c (hunk 7 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/__timedwait.c]: Unnecessary code size optimization change, reverted with commit 6ea3ad394b1b870f157be7162b7546f64a8896e3.

File: system/lib/libc/musl/src/thread/__timedwait.c (hunk 8 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/__timedwait.c]: Reverted with commit a9550ffb15068ee99de7b9db434b4b1f68383f8c. I'll make a new PR with this.

File: system/lib/libc/musl/src/thread/__wait.c (hunk 9 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/__wait.c]: Commit 190b6b93a41391f51b9a5701feeda123a1009ddd fixes this for the changed lines.

File: system/lib/pthread/library_pthread.c (hunk 2 above)
Comment (sbc100) [reply system/lib/pthread/library_pthread.c]: I think this camelCase exists here because it was ported from JS where we do use camelCase.    How about doing a naming-only PR.. either before or after this one is fine with.

**Comment by sbc100** [comment]
LGTM in general.

I have been thinking we can probable followup by removing a lot of this code... since emscripten_futex_wait itself already does time slicing on the main thread, I'm not sure its need at all in these places!

**Comment by sbc100** [comment]
@juj, what do you think about this?   If `emscripten_futex_wait` already does time slicing do you know why we are also doing it in all these higher level places?

**Comment by kleisauke** [comment]
I think it's still needed for cancellation points (i.e. for cases where the thread is cancelled in between). This is done in `__timedwait_cp`, `__wait`, `emscripten_thread_sleep`.  We might experiment with calling `emscripten_futex_wait` directly in `pthread_barrier_wait`.

**Comment by sbc100** [comment]
> I think it's still needed for cancellation points (i.e. for cases where the thread is cancelled in between). This is done in `__timedwait_cp`, `__wait` and `emscripten_thread_sleep`. We might experiment with calling `emscripten_futex_wait` directly in `pthread_barrier_wait`.

What if we adding a cancellation point to the `emscripten_futex_wait`?  Would that be a good idea anyway?

The alternative, would be to remove the time slicing from the emscripten_futex_wait loop.  I tried that here but decided it would be probably be better to keep this one and drop the others: https://github.com/emscripten-core/emscripten/pull/15975

File: system/lib/pthread/library_pthread.c [hunk 10]
```diff
@@ -99,23 +94,21 @@ void emscripten_thread_sleep(double msecs) {
...
-      emscripten_futex_wait(&dummyZeroAddress, 0, ms_to_sleep);
-    }
-    now = emscripten_get_now();
-  };
+    ms_to_sleep = target - emscripten_get_now();
+    if (ms_to_sleep < min_ms_slice_to_sleep)
+      continue;
+
+    emscripten_futex_wait(&dummyZeroAddress, 0,
+      ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep);
```
Comment (juj) [inline system/lib/pthread/library_pthread.c:110]: Here the code used to be of form
```c
emscripten_futex_wait(..., ms_to_sleep);
```
but it gets refactored to form
```c
emscripten_futex_wait(..., ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep);
```

Above in file `system/lib/libc/musl/src/thread/__wait.c` the code used to be

```c
emscripten_futex_wait(..., is_runtime_thread ? 1 : 100);
```
but it gets refactored to
```c
emscripten_futex_wait(..., max_ms_to_sleep);
```

The PR looks to me like it is massaging the code back and forth, I am not sure if I see the guiding idea here. (sorry!)

File: system/lib/libc/musl/src/thread/__timedwait.c [hunk 11]
```diff
@@ -75,18 +78,14 @@ int __timedwait_cp(volatile int *addr, int val,
...
-				r = ETIMEDOUT;
-				break;
-			}
-			if (waitMsecs > 100) waitMsecs = 100; // non-main threads can sleep in longer slices.
-			if (is_runtime_thread && waitMsecs > 1) waitMsecs = 1; // the runtime thread may need to run proxied calls, so sleep in very small slices to be responsive.
-			r = -emscripten_futex_wait((void*)addr, val, waitMsecs);
-		} while(r == ETIMEDOUT);
+
+			msecsToSleep = sleepUntilTime - emscripten_get_now();
+		} while (r == ETIMEDOUT && msecsToSleep > 0);
```
Comment (juj) [inline system/lib/libc/musl/src/thread/__timedwait.c:88]: This change has the subtle difference that before, if one had a wait time of zero, then the wait loop would process `emscripten_main_thread_process_queued_calls()` and exit, but after this change, there is now an unconditional call out to `emscripten_futex_wait()`. It is uncertain what that will do, it could yield back the remaining time slice at minimum, or it could do a full kernel wait that gets triggered via the general timeout mechanism, so performance could theoretically be affected.

**Review by juj — APPROVED** [review]
lgtm

**Comment by juj** [comment]
> What if we adding a cancellation point to the emscripten_futex_wait?

I think it would be best not to have emscripten_futex_wait depend on the pthreads API. That was probably the guiding principle - the Emscripten API resides on a lower level that is not tied to POSIX pthreads itself.

File: system/lib/pthread/library_pthread.c (hunk 10 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c:110]: Ah yes, it's a bit confusing. The difference between `system/lib/libc/musl/src/thread/__wait.c` and  `system/lib/pthread/library_pthread.c` is that the former needs to wait indefinitely in slices, so it's doing this:
```c
emscripten_futex_wait(..., max_ms_slice_to_sleep);
```
and the later must wait in slices for a certain time, so it does this:
```c
emscripten_futex_wait(..., MAX(ms_to_sleep, max_ms_slice_to_sleep));
```

I've renamed the `max_ms_to_sleep` variable to `max_ms_slice_to_sleep`  with commit 5b77bd909c717da9e11781a01e3d0a0a1b7c0039.

File: system/lib/pthread/library_pthread.c (hunk 2 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: Naming-only PR sounds good to me.

**Comment by kleisauke** [comment]
Test failures in https://circleci.com/gh/emscripten-core/emscripten/492229 looks real, I could reproduce that sometimes locally. It seems that `emscripten_main_thread_process_queued_calls()` must always called before `emscripten_futex_wait()` in order to fix that (see commit 9477d68ba5f2a6764faa0e9b4709b916f01dccbe).

File: system/lib/pthread/library_pthread.c (hunk 10 above)
Comment (sbc100) [reply system/lib/pthread/library_pthread.c:110]: I'd rather avoid the ternary operation here.  I think it makes the code easier to read if the calculation for how long to sleep is doing before the line where the sleep occurs.    It also make it easier for debugging since you have a name for the actual amount of time that is we are going to sleep for.

Don't you find "ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep" in addition to need to function call split over two lines harder to read?

Comment (kleisauke) [reply system/lib/pthread/library_pthread.c]: You're right, those long variable names make it hard to read when being used within a ternary operator. Split over 2 lines with commit f84985c239f0a0a089248e42f1a3ccebe430d1cf to improve readability.

**Comment by kleisauke** [comment]
Rebased to fix code size test failures in `test_minimal_runtime_code_size_hello_webgl{,2}_wasm2js` (commit f1fbd4c20a2419b38d965651f6a8e671fb87ba80).

**Comment by sbc100** [comment]
So is this change now completely non-functional?   How sure are you that there are no observable changes?  

Its not clear to me exactly what we gain from landing this PR?   Is the idea just to make variable call sites be more consistent? 
 Or is there a codesize or readability improvement here?

I'm not against landing this if it really is NFC.

**Comment by kleisauke** [comment]
> So is this change now completely non-functional? How sure are you that there are no observable changes?

I think all changes are completely non-functional expect for the `pthread_barrier_wait.c` change (which unifies the `emscripten_main_thread_process_queued_calls` and `emscripten_futex_wait` call order). That change looks like a bug fix, as the commit prior to 40064d65009c4282a7549097fb6c6e0f6dc661e4 where causing test failures.

> Its not clear to me exactly what we gain from landing this PR? Is the idea just to make variable call sites be more consistent?
> Or is there a codesize or readability improvement here?

You're right, the current state of this PR is that there is not much to gain other than readability and consistency. The reason for this PR was mainly the now-reverted 21c0823c419a15b607f383d733d7c77cc4bfa2e1 commit, which I'll open a PR for later.

> I'm not against landing this if it really is NFC.

I completely understand that, I just reverted the `pthread_barrier_wait.c` change with commit ecbd5fa2aee4ce4df1b6604af037b0cc23bd4bb6.

**Review by sbc100 — APPROVED** [review]

File: system/lib/pthread/library_pthread.c [hunk 12]
```diff
@@ -99,23 +94,21 @@ void emscripten_thread_sleep(double msecs) {
...
+      continue;
+    if (ms_to_sleep > max_ms_slice_to_sleep)
       ms_to_sleep = max_ms_slice_to_sleep;
-    }
-    if (ms_to_sleep >= min_ms_slice_to_sleep) {
-      emscripten_futex_wait(&dummyZeroAddress, 0, ms_to_sleep);
-    }
-    now = emscripten_get_now();
-  };
+    emscripten_futex_wait(&dummyZeroAddress, 0, ms_to_sleep);
```
[the remark this answers is not in the document]
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c:111]: Commit 7d74c5316dc8bd1b4e4df1eefb10cf375f73cb18 takes the elapsed time during `emscripten_futex_wait` into account.
